File: baekjoon/9주차/2294-동전-2/제문/2294.kt
package ndb.dp

import kotlin.math.min
import java.lang.StringBuilder

private val out = StringBuilder()
private val br = System.`in`.bufferedReader()

private fun readLine() = br.readLine()!!
private fun readInt() = br.readLine().toInt()
private fun readInts() = br.readLine().split(" ").map { it.toInt() }

fun main() {
    val (n, m) = readInts()
    val coins = IntArray(n) { readInt() }
    val dp = IntArray(m + 1) { 10001 }

    dp[0] = 0

    for (c in coins) {
        for (i in c..m) {
            dp[i] = min(dp[i], dp[i - c] + 1)
        }
    }

    println(if (dp[m] == 10001) -1 else dp[m])
}

File: baekjoon/9주차/2294-동전-2/우섭/2294.py
from sys import stdin
n, k = map(int, stdin.readline().split())
dp = [20000 for _ in range(k+1)]
coin = []
dp[0] = 0
for _ in range(n):
    coin.append(int(stdin.readline()))
for i in range(1, k+1):
    for j in coin:
        if i - j >= 0:
            dp[i] = min(dp[i], dp[i - j] + 1)
print(dp[k] if dp[k] != 20000 else -1)
